### ISSUE TITLE
Test dropdown value and select

### ISSUE THREAD

**Issue by senayar** [angular/protractor#1815]
Hi I have a simple test where I click on a dropdown button and select a value inside, the test work correctly in Chrome but with PhantomJS it's always failing with "No element found using locator": 

PhantomJS : 1.9.8
Protractor version : 1.61
Selenium 2.43.1

```javascript
  it('should verify than dropdown working', function () {
    browser.waitForAngular();
    var dropdownCategories = element(by.css('button.btn-select.dropdown-toggle'));
    expect(dropdownCategories.getText()).toBe('Toutes les catégories');
    dropdownCategories.click().then(function(){
      browser.waitForAngular();
      browser.sleep(1000).then(function(){
        element(by.css('ul.dropdown-menu > li:nth-child(2)')).click().then(function () {
          browser.waitForAngular();
          expect(dropdownCategories.getText()).toBe('Culture');
        });
      });
    });
  });
```
```javascript
1) News Page should verify than dropdown working
   Message:
     NoSuchElementError: No element found using locator: By.cssSelector("ul.dropdown-menu > li:nth-child(2)")
   Stacktrace:
     NoSuchElementError: No element found using locator: By.cssSelector("ul.dropdown-menu > li:nth-child(2)")
```

```html
<div class="table ng-scope">
  <div class="table-row">
    <form class="form-inline ng-pristine ng-valid">
      <div class="form-group">
        <label class="assistive-text">Trier par</label>
        <div class="dropdown open" on-toggle="toggled(open)">
          <button class="btn btn-select btn-default dropdown-toggle ng-binding" type="button" aria-haspopup="true" aria-expanded="true">
            Toutes les catégories
            <span class="arrow dropdown-menu-right">
              <svg class="icon icon-arrow-down-small" viewBox="0 0 50 32"><use xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="#icon-arrow-down-small"></use></svg>
            </span>
          </button>
          <ul class="dropdown-menu" role="menu" id="categoriesList">
            <li ng-click="changeCategory('all', 'Toutes les catégories')">
              <a>Toutes les catégories</a>
            </li>
            <!-- ngRepeat: category in categories --><li ng-repeat="category in categories" ng-click="changeCategory(category.name, category.label)" class="ng-scope">
              <a class="ng-binding">Culture</a>
            </li><!-- end ngRepeat: category in categories --><li ng-repeat="category in categories" ng-click="changeCategory(category.name, category.label)" class="ng-scope">
              <a class="ng-binding">Faits divers</a>
            </li><!-- end ngRepeat: category in categories --><li ng-repeat="category in categories" ng-click="changeCategory(category.name, category.label)" class="ng-scope">
              <a class="ng-binding">Football</a>
            </li><!-- end ngRepeat: category in categories --><li ng-repeat="category in categories" ng-click="changeCategory(category.name, category.label)" class="ng-scope">
              <a class="ng-binding">Insolite</a>
            </li><!-- end ngRepeat: category in categories --><li ng-repeat="category in categories" ng-click="changeCategory(category.name, category.label)" class="ng-scope">
              <a class="ng-binding">La Une</a>
            </li><!-- end ngRepeat: category in categories --><li ng-repeat="category in categories" ng-click="changeCategory(category.name, category.label)" class="ng-scope">
              <a class="ng-binding">Medecine</a>
            </li><!-- end ngRepeat: category in categories --><li ng-repeat="category in categories" ng-click="changeCategory(category.name, category.label)" class="ng-scope">
              <a class="ng-binding">People</a>
            </li><!-- end ngRepeat: category in categories --><li ng-repeat="category in categories" ng-click="changeCategory(category.name, category.label)" class="ng-scope">
              <a class="ng-binding">Politique</a>
            </li><!-- end ngRepeat: category in categories --><li ng-repeat="category in categories" ng-click="changeCategory(category.name, category.label)" class="ng-scope">
              <a class="ng-binding">Sports</a>
            </li><!-- end ngRepeat: category in categories --><li ng-repeat="category in categories" ng-click="changeCategory(category.name, category.label)" class="ng-scope">
              <a class="ng-binding">Économie &amp; Finance</a>
            </li><!-- end ngRepeat: category in categories -->
          </ul>
        </div>
      </div>
    </form>
    <div class="actions">
      <a class="button pull-right" ui-sref="root.news.category" href="/news/category">
        <svg class="icon icon-news-menu" viewBox="0 0 16 16"><use xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="#icon-news-menu"></use></svg>
      </a>
    </div>
  </div>
</div>
```
 same problem with this test : 

```javascript
  it('should click on the first topic in the list and display the news', function(){
    expect(element(by.css('div.help-row'))).not.toBe(null);
    browser.sleep(30000);
    element.all(by.repeater('news in filteredNews')).first().click().then(function(){
      browser.waitForAngular();
      browser.sleep(10000);
      expect(element(by.css('div.news-detail'))).not.toBe(null);
    });
  });
```

```
Failures:

  1) News Page should click on the first topic in the list and display the news
   Message:
     Error: Index out of bound. Trying to access element at index: 0, but there are only 0 elements that match locator by.repeater("news in filteredNews")
   Stacktrace:
     Error: Index out of bound. Trying to access element at index: 0, but there are only 0 elements that match locator by.repeater("news in filteredNews")

```
truncated HTML list (got more items inside)
```html
<ul class="news">
    <!-- ngRepeat: news in filteredNews = (newsList | filter:filterCategory) --><li ui-sref="root.news.show({id: news.id})" ng-repeat="news in filteredNews = (newsList | filter:filterCategory)" ng-click="saveNews(news, $index)" ng-class="{selected : news.selected == true}" class="ng-scope">
      <p class="ng-binding">
        Loi Macron: Manuel Valls accroît la pression sur les députés PS avant le vote <br>
        <small class="ng-binding">16.02.2015 à 13:28</small>
      </p>
    </li><!-- end ngRepeat: news in filteredNews = (newsList | filter:filterCategory) --><li ui-sref="root.news.show({id: news.id})" ng-repeat="news in filteredNews = (newsList | filter:filterCategory)" ng-click="saveNews(news, $index)" ng-class="{selected : news.selected == true}" class="ng-scope">
      <p class="ng-binding">
        Loi Macron: Manuel Valls accroît la pression sur les députés PS avant le vote <br>
        <small class="ng-binding">16.02.2015 à 13:28</small>
      </p>
    </li><!-- end ngRepeat: news in filteredNews = (newsList | filter:filterCategory) --><li ui-sref="root.news.show({id: news.id})" ng-repeat="news in filteredNews = (newsList | filter:filterCategory)" ng-click="saveNews(news, $index)" ng-class="{selected : news.selected == true}" class="ng-scope">
      <p class="ng-binding">
        Six nations: Pascal Papé en commission de discipline <br>
        <small class="ng-binding">16.02.2015 à 13:17</small>
      </p>
    </li><!-- end ngRepeat: news in filteredNews = (newsList | filter:filterCategory) --><li ui-sref="root.news.show({id: news.id})" ng-repeat="news in filteredNews = (newsList | filter:filterCategory)" ng-click="saveNews(news, $index)" ng-class="{selected : news.selected == true}" class="ng-scope">
      <p class="ng-binding">
        Mercedes va rappeler 127.000 véhicules en Chine <br>
        <small class="ng-binding">16.02.2015 à 13:06</small>
      </p>
    </li><!-- end ngRepeat: news in filteredNews = (newsList | filter:filterCategory) --><li ui-sref="root.news.show({id: news.id})" ng-repeat="news in filteredNews = (newsList | filter:filterCategory)" ng-click="saveNews(news, $index)" ng-class="{selected : news.selected == true}" class="ng-scope">
      <p class="ng-binding">
        Ligue des championse: carré de favoris et rêves de trouble-fête <br>
        <small class="ng-binding">16.02.2015 à 13:02</small>
      </p>
    </li><!-- end ngRepeat: news in filteredNews = (newsList | filter:filterCategory) --><li ui-sref="root.news.show({id: news.id})" ng-repeat="news in filteredNews = (newsList | filter:filterCategory)" ng-click="saveNews(news, $index)" ng-class="{selected : news.selected == true}" class="ng-scope">
      <p class="ng-binding">
        Ligue des championse: carré de favoris et rêves de trouble-fête <br>
        <small class="ng-binding">16.02.2015 à 13:02</small>
      </p>
    </li><!-- end ngRepeat: news in filteredNews = (newsList | filter:filterCategory) --><li ui-sref="root.news.show({id: news.id})" ng-repeat="news in filteredNews = (newsList | filter:filterCategory)" ng-click="saveNews(news, $index)" ng-class="{selected : news.selected == true}" class="ng-scope">
      <p class="ng-binding">
        Cyril Linette (Canal+) nommé directeur général de L'Equipe <br>
        <small class="ng-binding">16.02.2015 à 12:36</small>
      </p>
    </li><!-- end ngRepeat: news in filteredNews = (newsList | filter:filterCategory) -->
  </ul>

**Comment by nortain** [comment]
I am also seeing this kind of an error when using phantomjs to read the selected value of a dropdown.  I am fairly new to protractor but I was trying to test what the text value of the currently selected option is in a dropdown menu and have a test that passes in chrome but fails in phantomjs.
```
describe("testing differences between phantomjs and chrome", function() {
    beforeEach(function () {
        browser.get('/');
        reasonsForVisit = element(by.model('apptDetailCtrl.selectedReason'));
    });

    it('setting a value to the first dropdown menu on appt details', function () {
        selectDropdownbyNum(reasonsForVisit, 1);//sets the drop down to the first index which has the string text of Eye Infection/Problem

        expect(reasonsForVisit.$('option:checked').getText()).toEqual('Eye Infection/Problem');//this test passes in chrome but fails in phantomjs
    });

    var selectDropdownbyNum = function (element, optionNum) {
        if (optionNum) {
            var options = element.getWebElement().findElements(by.tagName('option'))
                .then(function (options) {
                    options[optionNum].click();
                });


        }
    };
});

```

While this passes in chrome the error message I see w/ phantomjs is as follows:

```
Message:
     NoSuchElementError: No element found using locator: By.cssSelector("option:checked")
   Stacktrace:
     NoSuchElementError: No element found using locator: By.cssSelector("option:checked")
    at Array.forEach (native)
    at runMicrotasksCallback (node.js:337:7)
```


In my protractor file I have the two different capabilities that I toggle between the two browsers.
```
...
capabilities: {
        'browserName': 'phantomjs',
        'phantomjs.binary.path': require('phantomjs').path,
        'phantomjs.ghostdriver.cli.args': ['--debug=true']

    },

    //capabilities: {
    //    'browserName': 'chrome',    
    //}
...
```

Since I have these tests wired up with livereload I'd much rather prefer using phantomjs to have them run in the background over chrome but I haven't figured out a way to actually accomplish this kind of a test with phantomjs.

The closest I came up with that passes in phantomjs was something like,
```
expect(reasonsForVisit.$$('option').get(1).getText()).toEqual('Eye Infection/Problem');
```
which isn't really accomplishing anything because I'm merely confirming that the first element of the dropdown menu has the text 'Eye Infection/Problem' but not that the dropdown's currently selected value is set to that.

For completeness here is the segment of my html:
```
<select id="reasonForVisit" class="form-control" ng-model="apptDetailCtrl.selectedReason" required
        ng-options="reason.id as reason.displayName for reason in apptDetailCtrl.reasonsForVisit"
        ng-change="apptDetailCtrl.changedReason()">
    <option value="">&lt;Select&gt;</option>
</select>
```
the reasonsForVisit being passed into ng-options is a collection being passed in via an external api.

**Comment by seminull** [comment]
I'm also experiencing this issue.

**Comment by tttp** [comment]
Same issue,

- PhantomJS : 1.9.8
- Protractor: 2.1.0
-  Selenium 2.45.0

The one line that works on chrome but fails in phantomjs:

     element(by.id(id)).$('option:checked').getText();

**Comment by KrekkieD** [comment]
Luckily the `option:checked` issue is fixed in phantomjs v2.0.0. Sadly, this version is not yet fully available for all platforms. It is on windows though, and can be installed through npm by running `npm install phantomjs2` as the official package (`phantomjs`) will not provide v2 until it is stable across the board.

**Comment by juliemr** [comment]
Phantom2 is out, closing this as stale.
